File: odroid/Command/src/bert/command/model/RobotCommandModel.java
/**
 * Copyright 2018-2019. Charles Coughlin. All Rights Reserved.
 *                 MIT License.
 */
package bert.command.model;

import java.nio.file.Path;
import java.util.logging.Logger;

import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import bert.share.message.HandlerType;
import bert.share.model.AbstractRobotModel;
import bert.share.model.ConfigurationConstants;
import bert.share.xml.XMLUtility;

/**
 *  This is the base class for a collection of models that keep basic configuration
 *  information, all reading from the same files. The information 
 */
public class RobotCommandModel extends AbstractRobotModel   {
	private static final String CLSS = "RobotCommandModel";
	private static final Logger LOGGER = Logger.getLogger(CLSS);
	private int blueserverPort = 11046;
	private String deviceUUID  = "";
	private String deviceMAC   = "";

	public RobotCommandModel(Path configPath) {
		super(configPath);
	}
    
	
	/**
	 *  Analyze the document and populate the model. 
	 */
	public void populate() {
		analyzeControllers();
		analyzeProperties();
		analyzeMotors();
	}
	public int getBlueserverPort() { return this.blueserverPort; }
	public String getDeviceMAC() { return deviceMAC; }
	/**
	 * @return bluetooth device service UUID as a String
	 */
	public String getDeviceUUID() { return deviceUUID; }

    // ================================ Auxiliary Methods  ===============================
	/**
	 * Search the XML configuration for the command controller. It's the only one we care about.
	 * If not found, the controller will be null.
	 */
	@Override
	public void analyzeControllers() {
		if( this.document!=null ) {
			String controllerName = "UNASSIGNED";
			NodeList elements = document.getElementsByTagName("controller");
			int count = elements.getLength();
			int index = 0;
			while(index<count) {
				Element controllerElement= (Element)(elements.item(index));
				controllerName = XMLUtility.attributeValue(controllerElement, "name");
				String type = XMLUtility.attributeValue(controllerElement, "type");
				if( type!=null && !type.isEmpty() &&
					type.equalsIgnoreCase(HandlerType.COMMAND.name()) ) {
					// Configure the socket - there should only be one.
					NodeList socketElements = controllerElement.getElementsByTagName("socket");
					int nSocket = socketElements.getLength();
					if( nSocket>0) {
						handlerTypes.put(controllerName,type.toUpperCase());
						for( int iSocket=0;iSocket<nSocket;iSocket++) {
							Element socketElement= (Element)(socketElements.item(iSocket));
							String socketMAC = XMLUtility.attributeValue(socketElement, "mac");
							String socketName = XMLUtility.attributeValue(socketElement, "name");
							String portName = XMLUtility.attributeValue(socketElement, "port");
							String socketType = XMLUtility.attributeValue(socketElement, "type");
							String socketUUID = XMLUtility.attributeValue(socketElement, "uuid");
							if( socketType.equalsIgnoreCase("bluetooth")) {
								deviceUUID = socketUUID;
								deviceMAC  = socketMAC;
							}
							else {
								sockets.put(controllerName,Integer.parseInt(portName));
							}
						}
					}
					break;
				}
				
				index++;
			}
			properties.put(ConfigurationConstants.PROPERTY_CONTROLLER_NAME, controllerName);
		}
	}
	/**
	 * Extend the default search for properties to convert the "blueserver" port to an int.
	 */
	@Override
	protected void analyzeProperties() {
		super.analyzeProperties();
		String port = properties.getProperty("blueserver");
		if( port!=null ) {
			try {
				this.blueserverPort = Integer.parseInt(properties.getProperty("blueserver"));
			}
			catch (NumberFormatException nfe) {
				LOGGER.warning(String.format("%s.analyzeProperties: Port for \"blueserver\" not a number (%s)",CLSS,nfe.getLocalizedMessage()));
			}
		}
		else {
			LOGGER.warning(String.format("%s.analyzeProperties: Port for \"blueserver\" missing in XML",CLSS));
		}
	}
}

File: Bert/command/src/main/kotlin/chuckcoughlin/bert/command/model/RobotCommandModel.kt
/**
 * Copyright 2018-2019. Charles Coughlin. All Rights Reserved.
 * MIT License.
 */
package chuckcoughlin.bert.command.model

import chuckcoughlin.bert.common.message.HandlerType
import chuckcoughlin.bert.common.model.AbstractRobotModel
import chuckcoughlin.bert.common.model.ConfigurationConstants
import chuckcoughlin.bert.common.util.XMLUtility
import org.w3c.dom.Element
import org.w3c.dom.NodeList
import java.nio.file.Path
import java.util.*
import java.util.logging.Logger

/**
 * This is the base class for a collection of models that keep basic configuration
 * information, all reading from the same files. The information
 */
class RobotCommandModel(configPath: Path?) : AbstractRobotModel(configPath) {
    var blueserverPort = 11046
        private set

    /**
     * @return bluetooth device service UUID as a String
     */
    var deviceUUID = ""
        private set
    var deviceMAC = ""
        private set

    /**
     * Analyze the document and populate the model.
     */
    fun populate() {
        analyzeControllers()
        analyzeProperties()
        analyzeMotors()
    }
    // ================================ Auxiliary Methods  ===============================
    /**
     * Search the XML configuration for the command controller. It's the only one we care about.
     * If not found, the controller will be null.
     */
    fun analyzeControllers() {
        if (this.document != null) {
            var controllerName = "UNASSIGNED"
            val elements: NodeList = document.getElementsByTagName("controller")
            val count = elements.length
            var index = 0
            while (index < count) {
                val controllerElement = elements.item(index) as Element
                controllerName = XMLUtility.attributeValue(controllerElement, "name")
                val type: String = XMLUtility.attributeValue(controllerElement, "type")
                if (type != null && !type.isEmpty() &&
                    type.equals(HandlerType.COMMAND.name(), ignoreCase = true)
                ) {
                    // Configure the socket - there should only be one.
                    val socketElements = controllerElement.getElementsByTagName("socket")
                    val nSocket = socketElements.length
                    if (nSocket > 0) {
                        handlerTypes.put(controllerName, type.uppercase(Locale.getDefault()))
                        for (iSocket in 0 until nSocket) {
                            val socketElement = socketElements.item(iSocket) as Element
                            val socketMAC: String = XMLUtility.attributeValue(socketElement, "mac")
                            val socketName: String = XMLUtility.attributeValue(socketElement, "name")
                            val portName: String = XMLUtility.attributeValue(socketElement, "port")
                            val socketType: String = XMLUtility.attributeValue(socketElement, "type")
                            val socketUUID: String = XMLUtility.attributeValue(socketElement, "uuid")
                            if (socketType.equals("bluetooth", ignoreCase = true)) {
                                deviceUUID = socketUUID
                                deviceMAC = socketMAC
                            } else {
                                sockets.put(controllerName, portName.toInt())
                            }
                        }
                    }
                    break
                }
                index++
            }
            properties.put(ConfigurationConstants.PROPERTY_CONTROLLER_NAME, controllerName)
        }
    }

    /**
     * Extend the default search for properties to convert the "blueserver" port to an int.
     */
    protected fun analyzeProperties() {
        super.analyzeProperties()
        val port: String = properties.getProperty("blueserver")
        if (port != null) {
            try {
                blueserverPort = properties.getProperty("blueserver").toInt()
            } catch (nfe: NumberFormatException) {
                LOGGER.warning(
                    String.format(
                        "%s.analyzeProperties: Port for \"blueserver\" not a number (%s)",
                        CLSS,
                        nfe.localizedMessage
                    )
                )
            }
        } else {
            LOGGER.warning(String.format("%s.analyzeProperties: Port for \"blueserver\" missing in XML", CLSS))
        }
    }

    companion object {
        private const val CLSS = "RobotCommandModel"
        private val LOGGER = Logger.getLogger(CLSS)
    }
}